### PR TITLE
AP_AerobridgeTrustedFlight: Add Aerobridge trusted flight subsystem

### PR DESCRIPTION
**Pull request by rhythmize**
This subsystem enables token verification before arming the vehicle. The token needs to be valid at the time of arming and issued by a valid issuer (currently `https://id.openskies.sh/`) failing which the vehicle will not arm.

This is an opt-in feature and can be enabled by providing `--trusted-flight-issuer` (JWT token issuer) and `--trusted-flight-root-certificate` (Root Certificate to establish root of trust for token verification) during `./waf configure`.

Helper scripts and README are added in `Tools/AerobridgeTrustedFlight_tools/`

### CODE REVIEW

**Comment by amilcarlucas** [comment]
Looks like the build size comparison did not work on this one:
```
Binary Name      Text [B]        Data [B]     BSS (B)        Total Flash Change [B] (%)      Flash Free After PR (B)
---------------  --------------  -----------  -------------  ----------------------------  -------------------------
antennatracker   -20 (-0.0015%)  0 (0.0000%)  -4 (-0.0015%)  -20 (-0.0015%)                                   608072
ardusub          -20 (-0.0013%)  0 (0.0000%)  -4 (-0.0015%)  -20 (-0.0013%)                                   385628
arducopter-heli  -20 (-0.0011%)  0 (0.0000%)  -4 (-0.0015%)  -20 (-0.0011%)                                   172008
blimp            -20 (-0.0015%)  0 (0.0000%)  -4 (-0.0015%)  -20 (-0.0015%)                                   664988
ardurover        -20 (-0.0012%)  0 (0.0000%)  -4 (-0.0015%)  -20 (-0.0012%)                                   342976
arduplane        -20 (-0.0011%)  0 (0.0000%)  -4 (-0.0015%)  -20 (-0.0011%)                                   192168
arducopter       -20 (-0.0011%)  0 (0.0000%)  -4 (-0.0015%)  -20 (-0.0011%)                                   1786[32]
```

**Comment by IamPete1** [comment]
I wonder if this should be a backend option for open drone ID, both can block arming. I guess this does not also have telemetry requirement?

**Comment by rhythmize** [comment]
Hi @amilcarlucas, I'm not sure I correctly follow the build size comparison part you mentioned. Is it some mandatory check which is supposed to run on every new features? Do I need to whitelist something somewhere in the code? Can you please elaborate a bit on this? Thanks!

**Comment by amilcarlucas** [comment]
We usually take a look at the memory cost of every new added feature.
Something went wrong your feature reduces memory cost.

**Comment by rhythmize** [comment]
Ah I see. Can you please share some links/ documentation with more info on this and how can I obtain this info locally? I'll try to investigate what's going wrong.

**Comment by rhythmize** [comment]
@amilcarlucas, I ran the pretty_diff_size script locally and got a diff of 4 bytes, which I can confirm is caused by refactoring [here](https://github.com/ArduPilot/ardupilot/pull/22097/files#diff-19590cf17b8265289b1ac103bbaf4f4e8ce4f3541a036524326a368a7c3702b7R1542). Although I could not get the 20 bytes diff which we see in the ci build. Is it possible to force run the checks again to be sure, it was not something transient?

And re the memory costs about the new feature, it won't show up in the automated checks because this is an opt-in feature and is disabled by default. Additional details are added in the readme [here](https://github.com/ArduPilot/ardupilot/pull/22097/files#diff-3744fbf502870dc260e755f43c74b77e1c8bea427da629ff4b0dfa2bce3fd1d8)

File: .gitmodules [hunk 1]
```diff
@@ -34,3 +34,6 @@
 [submodule "modules/DroneCAN/libcanard"]
 	path = modules/DroneCAN/libcanard
 	url = https://github.com/DroneCAN/libcanard
+[submodule "modules/trusted_flight"]
+	path = modules/trusted_flight
+	url = https://github.com/GlitchedPolygons/l8w8jwt.git
```
Comment (tridge) [inline .gitmodules:39]: needs to be forked into ArduPilot github organisation

File: Tools/AerobridgeTrustedFlight_tools/1_pre_flight_checks_load_operation.py [hunk 2]
```diff
@@ -0,0 +1,259 @@
+import asyncio
```
Comment (tridge) [inline Tools/AerobridgeTrustedFlight_tools/1_pre_flight_checks_load_operation.py:1]: needs some comments in the python explaining what it does, also needs to have install script to install needed modules

File: libraries/AP_AerobridgeTrustedFlight/AP_AerobridgeTrustedFlight.cpp [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+
+    return read_file(public_key_path, public_key) && read_file(token_path, token) && verify_token(public_key, token);
+}
+
+bool AP_AerobridgeTrustedFlight::read_file(std::string &filename, std::string &filecontent)
+{
+    FileData *filedata = AP::FS().load_file(get_filepath(filename).c_str());
+
+    if (filedata == nullptr) {
+        hal.console->printf("Cannot load file %s\n", filename.c_str());
```
Comment (tridge) [inline libraries/AP_AerobridgeTrustedFlight/AP_AerobridgeTrustedFlight.cpp:51]: most users won't see these printf() calls, ise GCS_SEND_TEXT()

File: libraries/AP_AerobridgeTrustedFlight/AP_AerobridgeTrustedFlight.cpp [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+
+    enum l8w8jwt_validation_result validation_result;
+    struct l8w8jwt_claim claims;
+    struct l8w8jwt_claim* ref = &claims;
+    size_t claims_count;
+
+    int decode_result = l8w8jwt_decode(&params, &validation_result, &ref, &claims_count);
+
+    if (decode_result == L8W8JWT_DECODE_FAILED_INVALID_TOKEN_FORMAT) {
+        hal.console->printf("Aerobridge Trusted Flight: Token format invalid\n");
```
Comment (tridge) [inline libraries/AP_AerobridgeTrustedFlight/AP_AerobridgeTrustedFlight.cpp:88]: printf not seen, same for other printf calls later

File: libraries/AP_Arming/AP_Arming.cpp [hunk 5]
```diff
@@ -1535,14 +1539,17 @@ bool AP_Arming::arm(AP_Arming::Method method, const bool do_arming_checks)
 
     running_arming_checks = true;  // so we show Arm: rather than Disarm: in messages
 
-    if ((!do_arming_checks && mandatory_checks(true)) || (pre_arm_checks(true) && arm_checks(method))) {
-        armed = true;
+    armed = (!do_arming_checks && mandatory_checks(true)) || (pre_arm_checks(true) && arm_checks(method));
 
+#ifdef HAL_Aerobridge_Trusted_Flight
+    armed &= AP::aerobridge_trusted_flight().is_valid();
```
Comment (tridge) [inline libraries/AP_Arming/AP_Arming.cpp:1545]: need to test that this can't be bypassed with a forced arm

File: libraries/AP_AerobridgeTrustedFlight/AP_AerobridgeTrustedFlight.h [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    bool read_file(std::string &filename, std::string &filecontent);
+    bool verify_token(std::string &key, std::string &token);
+    std::string get_filepath(std::string &filename);
+
+    struct l8w8jwt_decoding_params params;
+    static AP_AerobridgeTrustedFlight *_singleton;
+    std::string public_key_path;
+    std::string token_path;
+    std::string issuer = "https://id.openskies.sh/";
+    std::string public_key_filename = "trusted_flight/auth_server_public_key.pem";
```
Comment (tridge) [inline libraries/AP_AerobridgeTrustedFlight/AP_AerobridgeTrustedFlight.h:46]: this must be in ROMFS or this whole system is totally insecure

**Review by tridge — CHANGES_REQUESTED** [review]

File: libraries/AP_Arming/AP_Arming.cpp [hunk 7]
```diff
@@ -51,6 +51,10 @@
 #include <AP_OpenDroneID/AP_OpenDroneID.h>
 #include <AP_SerialManager/AP_SerialManager.h>
 
+#ifdef HAL_Aerobridge_Trusted_Flight
```
[the remark this answers is not in the document]
Comment (hendjoshsr71) [reply libraries/AP_Arming/AP_Arming.cpp:54]: I think Peter meant `AP_AEROBRIDGE_TRUSTED_FLIGHT_ENABLED`

File: .gitignore [hunk 8]
```diff
@@ -152,3 +152,4 @@ dumpstack_*out
 build.tmp.binaries/
 tasklist.json
 modules/esp_idf
+**/trusted_flight/*
```
Comment (bugobliterator) [inline .gitignore:155]: new line after this

File: .gitmodules (hunk 1 above)
Comment (bugobliterator) [reply .gitmodules:44]: If it gets approved I will help creating a fork in ardupilot

File: Tools/AerobridgeTrustedFlight_tools/.env.sample [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+AEROBRIDGE_URL=https://aerobridgetestflight.herokuapp.com/
+AEROBRIDGE_CLIENT_ID=__CLIENT_ID__
+AEROBRIDGE_CLIENT_SECRET=__CLIENT_SECRET__
+
+# For submitting telemetry data to Flight Blender
+BLENDER_WRITE_CLIENT_ID=__BLENDER_CLIENT_ID__
+BLENDER_WRITE_CLIENT_SECRET=__BLENDER_CLIENT_SECRET__
+BLENDER_AUDIENCE=__BLENDER_AUDIENCE__
+BLENDER_WRITE_SCOPE=__BLENDER_WRITE_SCOPE__
+BLENDER_RID_FQDN=__BLENDER_RID_FQDN__
```
Comment (bugobliterator) [inline Tools/AerobridgeTrustedFlight_tools/.env.sample]: new line

File: libraries/AP_AerobridgeTrustedFlight/AP_AerobridgeTrustedFlight.cpp [hunk 10]
```diff
@@ -0,0 +1,130 @@
+/*
```
Comment (bugobliterator) [inline libraries/AP_AerobridgeTrustedFlight/AP_AerobridgeTrustedFlight.cpp:1]: ensure that new cpp files have gone through astyle check

File: wscript [hunk 11]
```diff
@@ -440,7 +431,7 @@ def configure(cfg):
     cfg.env.BOARD = cfg.options.board
     cfg.env.DEBUG = cfg.options.debug
     cfg.env.COVERAGE = cfg.options.coverage
-    cfg.env.FORCE32BIT = cfg.options.force_32bit
+    cfg.env.SITL32BIT = cfg.options.sitl_32bit
```
Comment (bugobliterator) [inline wscript]: please double check where and why these changes are coming from, they shouldn't be here I think

File: wscript [hunk 12]
```diff
@@ -666,20 +652,19 @@ def _build_dynamic_sources(bld):
 
     if (bld.get_board().with_can or bld.env.HAL_NUM_CAN_IFACES) and not bld.env.AP_PERIPH:
         bld(
-            features='dronecangen',
-            source=bld.srcnode.ant_glob('modules/DroneCAN/DSDL/* libraries/AP_DroneCAN/dsdl/*', dir=True, src=False),
-            output_dir='modules/DroneCAN/libcanard/dsdlc_generated/',
-            name='dronecan',
+            features='uavcangen',
```
Comment (bugobliterator) [inline wscript]: I think these are spurious changes. please review

**Review by bugobliterator — CHANGES_REQUESTED** [review]
Please split the changes into different commits if they impact different folders. We have a system of using <folder_name>:<commit_message>. Changes should be limited to that folder only.

File: wscript (hunk 11 above)
Comment (rhythmize) [reply wscript]: cleaned up. marked as outdated now

File: wscript (hunk 12 above)
Comment (rhythmize) [reply wscript]: cleaned up. marked as outdated now

File: libraries/AP_AerobridgeTrustedFlight/AP_AerobridgeTrustedFlight.cpp (hunk 10 above)
Comment (rhythmize) [reply libraries/AP_AerobridgeTrustedFlight/AP_AerobridgeTrustedFlight.cpp:1]: done!

File: Tools/AerobridgeTrustedFlight_tools/.env.sample (hunk 9 above)
Comment (rhythmize) [reply Tools/AerobridgeTrustedFlight_tools/.env.sample]: added.

File: .gitignore (hunk 8 above)
Comment (rhythmize) [reply .gitignore]: done!

File: libraries/AP_Arming/AP_Arming.cpp [hunk 13]
```diff
@@ -1588,16 +1592,18 @@ bool AP_Arming::arm(AP_Arming::Method method, const bool do_arming_checks)
 
     running_arming_checks = true;  // so we show Arm: rather than Disarm: in messages
 
-    if ((!do_arming_checks && mandatory_checks(true)) || (pre_arm_checks(true) && arm_checks(method))) {
-        armed = true;
+    armed = (!do_arming_checks && mandatory_checks(true)) || (pre_arm_checks(true) && arm_checks(method));
 
         _last_arm_method = method;
+#ifdef AP_AEROBRIDGE_TRUSTED_FLIGHT_ENABLED
+    armed &= AP::aerobridge_trusted_flight().is_trusted();
```
Comment (tridge) [inline libraries/AP_Arming/AP_Arming.cpp:1599]: should be in mandatory_checks(), alongside opendroneid checks

File: Tools/ardupilotwaf/boards.py [hunk 14]
```diff
@@ -448,7 +448,41 @@ def configure_env(self, cfg, env):
                 CANARD_ENABLE_DEADLINE = 1,
             )
 
+        if cfg.options.trusted_flight_issuer and cfg.options.trusted_flight_root_certificate:
+            # prepare a temp file to embed issuer string into ROMFS
+            with open('/tmp/trusted_flight_issuer.tmp', 'w') as f:
```
Comment (tridge) [inline Tools/ardupilotwaf/boards.py:453]: should use the python tmpfile capability, both for security and cleanup

File: libraries/AP_AerobridgeTrustedFlight/AP_AerobridgeTrustedFlight.cpp [hunk 15]
```diff
@@ -0,0 +1,193 @@
...
+{
+    if (_singleton != nullptr) {
+        AP_HAL::panic("Too many AerobridgeTrustedFlight modules");
+        return;
+    }
+
+    _singleton = this;
+}
+
+bool AP_AerobridgeTrustedFlight::is_trusted()
```
Comment (tridge) [inline libraries/AP_AerobridgeTrustedFlight/AP_AerobridgeTrustedFlight.cpp:34]: should have comment explaining what this function does

File: Tools/ardupilotwaf/boards.py (hunk 14 above)
Comment (tridge) [reply Tools/ardupilotwaf/boards.py:453]: or better putting the file in the build tree? Best not to create files outside of buildroot

**Review by tridge — CHANGES_REQUESTED** [review]
we need this to be built for a STM32 target (eg. CubeOrange) in CI, perhaps create a CubeOrange-AeroBridge target which embeds dummy certificates in the hwdef directory
We should also test that it works in SITL, with a proper arming check

File: .gitmodules [hunk 16]
```diff
@@ -39,3 +39,7 @@
 	path = modules/Micro-CDR
 	url = https://github.com/ardupilot/Micro-CDR.git
 	branch = master
+[submodule "modules/trusted_flight"]
+	path = modules/trusted_flight
```
Comment (tridge) [inline .gitmodules:43]: this should be modules/l8w8jwt

File: libraries/AP_AerobridgeTrustedFlight/AP_AerobridgeTrustedFlight.cpp (hunk 15 above)
Comment (tridge) [inline libraries/AP_AerobridgeTrustedFlight/AP_AerobridgeTrustedFlight.cpp:34]: this should look like this: ```
bool AP_AerobridgeTrustedFlight::is_trusted(size_t buflen, char *buffer)
{
  if (!good) {
  hal.util->snprintf(buffer, buflen, "No good - try again tomorrow");
}
}

Comment (rhythmize) [reply libraries/AP_AerobridgeTrustedFlight/AP_AerobridgeTrustedFlight.cpp:34]: @tridge What would be the best place to write debug/ info logs, if any? Or all the logs should go to the same buffer for displaying error messages? Also, since this is a one time pre-arm check should we log the success message for visibility or ignore it?

File: Tools/ardupilotwaf/boards.py (hunk 14 above)
Comment (rhythmize) [reply Tools/ardupilotwaf/boards.py]: Can't use tempfile because it's cleaned up as soon as fd is closed. Moved it to buildroot instead

File: libraries/AP_AerobridgeTrustedFlight/AP_AerobridgeTrustedFlight.cpp (hunk 15 above)
Comment (rhythmize) [reply libraries/AP_AerobridgeTrustedFlight/AP_AerobridgeTrustedFlight.cpp]: resolved

Comment (rhythmize) [reply libraries/AP_AerobridgeTrustedFlight/AP_AerobridgeTrustedFlight.cpp]: resolved

File: Tools/scripts/AP_AerobridgeTrustedFlight/requirements.txt [hunk 17]
```diff
@@ -0,0 +1,6 @@
+mavsdk==0.22.3
+requests==2.27.1
+python-dotenv==0.19.2
+pyjwt==2.3.0
+cryptography==36.0.1
+arrow==1.2.2
```
Comment (MattKear) [inline Tools/scripts/AP_AerobridgeTrustedFlight/requirements.txt:6]: need to add an additional line return

File: Tools/scripts/AP_AerobridgeTrustedFlight/utils/aerobridgetools.py [hunk 18]
```diff
@@ -0,0 +1,255 @@
...
+		
+		
+		securl = self.aerobridge_url + 'registry/aircraft/firmware/' + registered_flight_module_id
+		if self.token:
+			headers = {'Authorization': 'Bearer '+ self.token , 'content-type': 'application/json'}
+		else:
+			headers = {}
+		
+		r = self.session.get(securl, headers= headers)
+		return r
```
Comment (MattKear) [inline Tools/scripts/AP_AerobridgeTrustedFlight/utils/aerobridgetools.py:255]: need to add an additional line return

File: Tools/scripts/AP_AerobridgeTrustedFlight/utils/data_definitions.py [hunk 19]
```diff
@@ -0,0 +1,18 @@
...
+class FirmwareVersionAndHash:
+    '''A class to hold information about Firmware version and hash from the board'''
+    flight_sw_version: str
+    flight_sw_git_hash: str
+
+@dataclass
+class HardwareUID:
+    '''A class to hold information about Hardware version of the drone'''
+    hardware_uid : str
+
```
Comment (MattKear) [inline Tools/scripts/AP_AerobridgeTrustedFlight/utils/data_definitions.py:18]: need to add an additional line return

File: Tools/scripts/AP_AerobridgeTrustedFlight/utils/exceptions.py [hunk 20]
```diff
@@ -0,0 +1,2 @@
+class ExpiredCertificateException(Exception):
+    pass
```
Comment (MattKear) [inline Tools/scripts/AP_AerobridgeTrustedFlight/utils/exceptions.py:2]: need to add an additional line return

File: Tools/scripts/AP_AerobridgeTrustedFlight/requirements.txt (hunk 17 above)
Comment (rhythmize) [reply Tools/scripts/AP_AerobridgeTrustedFlight/requirements.txt]: done

File: Tools/scripts/AP_AerobridgeTrustedFlight/utils/aerobridgetools.py (hunk 18 above)
Comment (rhythmize) [reply Tools/scripts/AP_AerobridgeTrustedFlight/utils/aerobridgetools.py]: done

File: Tools/scripts/AP_AerobridgeTrustedFlight/utils/data_definitions.py (hunk 19 above)
Comment (rhythmize) [reply Tools/scripts/AP_AerobridgeTrustedFlight/utils/data_definitions.py]: done

File: Tools/scripts/AP_AerobridgeTrustedFlight/utils/exceptions.py (hunk 20 above)
Comment (rhythmize) [reply Tools/scripts/AP_AerobridgeTrustedFlight/utils/exceptions.py]: done

File: libraries/AP_Arming/AP_Arming.cpp [hunk 21]
```diff
@@ -1745,11 +1762,14 @@ bool AP_Arming::arm(AP_Arming::Method method, const bool do_arming_checks)
 
     running_arming_checks = true;  // so we show Arm: rather than Disarm: in messages
 
-    if ((!do_arming_checks && mandatory_checks(true)) || (pre_arm_checks(true) && arm_checks(method))) {
-        armed = true;
+    armed = (!do_arming_checks && mandatory_checks(true)) || (pre_arm_checks(true) && arm_checks(method));
```
Comment (peterbarker) [inline libraries/AP_Arming/AP_Arming.cpp:1765]: Should just be a mandatory check.

**Review by peterbarker — CHANGES_REQUESTED** [review]
Needs a thorough review.

Please put testing information in your description.  What testing's been done

Please put in a "limits of what this protection offers".  e.g. the hardware must be secure for any time-based access control to be used.

Want Valgrind testing results based on a good autotest suite.

What is the flash cost of this feature?

Can we just use MonoCypher to sign the missions and arming authorities and whatnot. so we don't need the mbedtls thing?  The jwt tokens could be handled by the GCS layer.

**Review by tridge — CHANGES_REQUESTED** [review]

**Comment by rhythmize** [comment]
Current flash cost for this feature is additional ~119 KB
Without trusted flight
Without Trusted Flight

With trusted flight
With Trusted Flight

**Comment by peterbarker** [comment]
Replaced by https://github.com/ArduPilot/ardupilot/pull/28715
